functional/stateful-types.md

```diff
@@ -0,0 +1,250 @@
+
+Suppose you are designing an interpreted language runtime, which requires a JIT
+compiler in order to process. Many Rust novices coming from other languages
+would implement it this way:
+
+```rust
+#[derive(Debug)]
+pub enum CompileError {
+    // error type, std::error::Error impl skipped for brevity
+}
+
+#[derive(Debug)]
+pub enum ExecError {
+    // error type, std::error::Error impl skipped for brevity
+}
+
+pub type ExecutionResult = Result<(), ExecError>;
+
+pub type CompileResult = Result<(), CompileError>;
+
+pub enum Param {
+    GCSize(usize),
+    MaxStack(usize),
+    // more parameters...
+}
+
+#[derive(Default)]
+pub struct Interpreter {
+    // state and execution data
+}
+
+impl Interpreter {
+    /// Set a system prameter.
+    ///
+    /// # Panics
+    /// Will panic if called after a script is executed.
+    ///
+    pub fn set_param(&mut self, p: Param) {
+        /* update the state based on the parameter */
+    }
+
+    /// Set up a new execution environment.
+    ///
+    /// # Panics
+    /// Will panic if called more than once or after a script has been compiled.
+    ///
+    pub fn init(&mut self) {
+        /* Create the heap, prepare for use... */
+    }
+
+    /// Load and compile a script. Returns any errors encountered.
+    ///
+    pub fn compile_script(&mut self, script: &str) -> CompileResult {
+        /* actually do the compile... */
+        Ok(())
+    }
+
+    /// Execute all scripts added. Returns any errors encountered.
+    ///
+    /// # Panics
+    /// Will panic if zero scripts have been compiled.
+    ///
+    pub fn exec(&mut self) -> ExecutionResult {
+        /* actually run the script... */
+        Ok(())
+    }
+}
+
+fn main() {
+    let mut interp = Interpreter::default();
+    interp.set_param(Param::GCSize(1024 * 1024 * 1024));
+    interp.init();
+    interp.compile_script("print('2 + 2')").unwrap();
+    interp.exec().unwrap();
+}
+```
+
+Why those chances to panic? Because there are *state invariants* here:
+
+1. `set_param` can only be called in an "initial" state.
+2. `init` must be called before any scripts are compiled.
+3. `exec` can only be called in a "loaded" or "ready" state.
+4. `init` must be called *exactly once*.
+
+It would be possible to add these to the `Error` types instead of panicking.
+However, this solution is suboptimal. Not only would users of the code
+correctly have to `unwrap()` a result all the time, but the requirement to call
+init exactly once is still unenforcable across an entire program without
+additional state (like `called = true` in the struct).
+
+What would really be neat is if it were possible to create a compile-time error
+if it were misused. After all, every user's program contains the invalid call
+order in the logic itself.
+
+In Rust, this is actually possible! The solution is to *change the type* in
+order to enforce the invariants.
+
+How? With a private generic parameter.
+
+Here is what that looks like:
+
+```rust,ignore
```

Comment:
Could be nice to have a working example here, indeed. So people can copy and paste it and try it out/play around with it.

Reply:
I only did that to avoid repeating all the typing. I can easily put that back in and make it work.

Reply:
Hmm, this is turning out to be trickier than I thought. I may have to refactor the example to make it simple enough. That will take me a bit.

Reply:
If you want to you can do that in another PR as well. It would be nice to have a working example that is checked, but not always necessary. Also this is quite complex so I'm fine with just refactoring it at a later stage, maybe even by someone else. @MarcoIeni thoughts?

Reply:
It's up to jhwgh. If you want you can do it in another PR :)